### PR TITLE
Add a switch argument to opam when context is not default

### PR DESCRIPTION
**Pull request by tmattio**
This PR adds a `switch` argument to the `opam install` command in the missing external dependencies hint.

Previously, if a library was missing in one of the context, the hint looked like this:
```
Hint: try:
  opam install astring
```

It now looks like this if the context is not "default":
```
Hint: try:
  opam install --switch=4.07.1 astring
```

### CODE REVIEW

**Comment by nojb** [comment]
The names of `dune` contexts do not coincide in general with `opam` switches, so not sure this change can produce non-sensical output in general.

**Comment by tmattio** [comment]
Thanks @nojb! I've updated the PR to use the switch name instead of the context name

File: bin/external_lib_deps.ml [hunk 1]
```diff
@@ -46,18 +46,31 @@ let all_lib_deps ~request =
...
+
 let run ~lib_deps ~by_dir ~setup ~only_missing ~sexp =
   Dune_engine.Context_name.Map.foldi lib_deps ~init:false
     ~f:(fun context_name lib_deps_by_dir acc ->
       let lib_deps =
         Path.Source.Map.values lib_deps_by_dir
         |> List.fold_left ~init:Lib_name.Map.empty ~f:Lib_deps_info.merge
       in
-      let internals =
+      let scontext =
```
Comment (nojb) [inline bin/external_lib_deps.ml:64]:
```suggestion
      let sctx =
```
This is the name usually used in the codebase.

File: bin/external_lib_deps.ml [hunk 2]
```diff
@@ -46,18 +46,31 @@ let all_lib_deps ~request =
...
         Dune_engine.Context_name.Map.find_exn setup.Import.Main.scontexts
           context_name
-        |> Super_context.internal_lib_names
       in
+      let opam_switch_name =
```
Comment (nojb) [inline bin/external_lib_deps.ml:68]:
```suggestion
      let switch_name =
```
this way you can use punning a few lines below.

File: bin/external_lib_deps.ml [hunk 3]
```diff
@@ -46,18 +46,31 @@ let all_lib_deps ~request =
...
+        match Super_context.context scontext |> fun ctx -> ctx.Context.kind with
```
Comment (nojb) [inline bin/external_lib_deps.ml:69]:
```suggestion
        match (Super_context.context scontext).Context.kind with
```
a bit more concise.

**Review by nojb — COMMENTED** [review]
Looks good, I left some style comments.

**Comment by tmattio** [comment]
@nojb I addressed the reviews - let me know if there's anything else, thanks 🙂

File: CHANGES.md [hunk 4]
```diff
@@ -74,6 +74,8 @@ Unreleased
 
 - Tweak auto-formatting of `dune` files to improve readability. (#3928, @nojb)
 
+- Add a switch argument to opam when context is not default. (#3951m @tmattio)
```
Comment (nojb) [inline CHANGES.md:77]:
```suggestion
- Add a switch argument to opam when context is not default. (#3951, @tmattio)
```

**Review by nojb — COMMENTED** [review]
LGTM

Let's wait and see if the CI passes; It may be that some tests need updating.

**Comment by tmattio** [comment]
> It may be that some tests need updating

It shouldn't, I've run the tests locally, but let's see!

File: bin/external_lib_deps.ml [hunk 5]
```diff
@@ -46,18 +46,31 @@ let all_lib_deps ~request =
   |> Context_name.Map.map
        ~f:(Path.Source.Map.of_list_reduce ~f:Lib_deps_info.merge)
 
+let populate_opam_command ?switch_name packages =
```
Comment (nojb) [inline bin/external_lib_deps.ml:49]: Apologies for the nitpicking, but I would rename this to `opam_install_command` instead. "Populate" is a bit confusing IMHO.

**Review by nojb — APPROVED** [review]
This is good to go, modulo a tiny naming suggestion.

File: bin/external_lib_deps.ml (hunk 5 above)
Comment (tmattio) [reply bin/external_lib_deps.ml]: Done!

**Comment by nojb** [comment]
Merged, thanks.

**Pull request merged**